File: 백준/Gold/2470. 두 용액/두 용액.py
import sys
from bisect import bisect_left,bisect_right
for test_case in range(1):
    n = int(sys.stdin.readline())
    arr = list(map(int, sys.stdin.readline().split()))
    arr.sort()

    min_total = float('inf')
    result = []
    for i in range(n):
        num = arr[i]
        idx = bisect_left(arr, num*-1)

        cases = [idx, idx + 1, idx - 1]
        min_num = float('inf')
        tmp = -1
        for case in cases:
            if 0 <= case < n:
                if min_num > abs(num + arr[case]):
                    min_num = abs(num + arr[case])
                    tmp = case
        if i == tmp:
            continue
        if min_total > min_num:
            min_total = min_num
            result = [num, arr[tmp]]

    result.sort()
    print(' '.join(map(str, result)))

File: 백준/Gold/2470. 두 용액/두 용액.java

import java.util.*;
import java.util.stream.Collectors;
import java.io.*;

public class Main
{
	static long[] num;
	static int n;
	public static void main(String args[]) throws Exception
	{
		//BufferedReader br = new BufferedReader(new FileReader("./input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));	
		
		n = Integer.parseInt(br.readLine());
		num = Arrays.stream(br.readLine().split(" ")).mapToLong(Long::parseLong).toArray();
		Arrays.sort(num);
		Long gap = Long.MAX_VALUE;
		Long[] ans = new Long[2]; 
		for (int i = 0; i < n; i++) {
			int idx = binary_search(num[i]*-1);
			int[] cases = {idx -1, idx, idx + 1};
			
			for (int c : cases) {
				if (i == c || c < 0 || c >= n) continue;
				long tmp = Math.abs(num[i] + num[c]);
				if (gap > tmp) {
					gap = tmp;
					Long[] a = {num[i], num[c]};
					ans = a;
				}
				
			}
		}
		Arrays.sort(ans);
		String a = "";
		for (Long an : ans) {
			a += Long.toString(an) + " ";
		}
		System.out.println(a);
    }
	
	static int binary_search(long target) {
		int start = 0;
		int end = n - 1;
		
		while (start <= end) {
			int mid = (start + end) / 2;
			
			if (num[mid] > target) {
				end = mid - 1;
			}else {
				start = mid + 1;
			}
		}
		
		return end;
	}
}